File: lahetysrajapinta/src/main/scala/fi/oph/viestinvalitys/vastaanotto/model/LahetysValidator.scala
package fi.oph.viestinvalitys.vastaanotto.model

import fi.oph.viestinvalitys.vastaanotto.model.Lahetys.*
import fi.oph.viestinvalitys.vastaanotto.model.LahetysImpl.*
import org.apache.commons.validator.routines.EmailValidator

import java.util.{List, Optional}
import scala.jdk.CollectionConverters.*
import scala.util.matching.Regex

/**
 * Validoi järjestelmään syötetyn lähetyksen kentät. Validaattorin virheilmoitukset eivät saa sisältää sensitiivistä
 * tietoa koska ne menevät mm. lokeille.
 */
object LahetysValidator:

  final val VALIDATION_OPH_OID_PREFIX                 = "1.2.246.562"
  final val VALIDATION_OPH_DOMAIN                     = "@opintopolku.fi"

  final val VALIDATION_OTSIKKO_TYHJA                  = "otsikko: Kenttä on pakollinen"
  final val VALIDATION_OTSIKKO_LIIAN_PITKA            = "otsikko: Otsikko ei voi pidempi kuin " + OTSIKKO_MAX_PITUUS + " merkkiä"

  final val VALIDATION_LAHETTAVA_PALVELU_TYHJA        = "lahettavaPalvelu: Kenttä on pakollinen"
  final val VALIDATION_LAHETTAVA_PALVELU_LIIAN_PITKA  = "lahettavaPalvelu: Kentän pituus voi olla korkeintaan " + LAHETTAVAPALVELU_MAX_PITUUS + " merkkiä"
  final val VALIDATION_LAHETTAVA_PALVELU_INVALID      = "lahettavaPalvelu: Arvo ei ole validi käännösavain"

  final val VALIDATION_LAHETTAJAN_OID_INVALID         = "lähettäjänOid: Oid ei ole validi (1.2.246.562-alkuinen) oph-oid"

  final val VALIDATION_LAHETTAJAN_OID_PITUUS          = "lähettäjänOid-kentän suurin sallittu pituus on " + VIRKAILIJAN_OID_MAX_PITUUS + " merkkiä"

  final val VALIDATION_LAHETTAJA_TYHJA                = "lähettäjä: Kenttä on pakollinen"
  final val VALIDATION_LAHETTAJA_NIMI_LIIAN_PITKA     = "lähettäjä: nimi-kenttä voi maksimissaan olla " + LAHETTAJA_NIMI_MAX_PITUUS + " merkkiä pitkä"
  final val VALIDATION_LAHETTAJAN_OSOITE_TYHJA        = "lähettäjä: Lähettäjän sähköpostiosoite -kenttä on pakollinen"
  final val VALIDATION_LAHETTAJAN_OSOITE_INVALID      = "lähettäjä: Lähettäjän sähköpostiosoite ei ole validi sähköpostiosoite"
  final val VALIDATION_LAHETTAJAN_OSOITE_DOMAIN       = "lähettäjä: Lähettäjän sähköpostiosoite ei ole opintopolku.fi -domainissa"

  final val VALIDATION_REPLYTO_INVALID                = "replyTo: arvo ei ole validi sähköpostiosoite"

  final val VALIDATION_PRIORITEETTI                   = "prioriteetti: Prioriteetti täytyy olla joko \"" + LAHETYS_PRIORITEETTI_NORMAALI + "\" tai \"" + LAHETYS_PRIORITEETTI_KORKEA + "\""

  final val VALIDATION_SAILYTYSAIKA_TYHJA             = "sailytysAika: Kenttä on pakollinen"
  final val VALIDATION_SAILYTYSAIKA                   = "sailytysAika: Säilytysajan tulee olla " + SAILYTYSAIKA_MIN_PITUUS + "-" + SAILYTYSAIKA_MAX_PITUUS + " päivää"

  def validateOtsikko(otsikko: Optional[String]): Set[String] =
    if (otsikko.isEmpty || otsikko.get.length == 0)
      Set(VALIDATION_OTSIKKO_TYHJA)
    else if (otsikko.get.length > OTSIKKO_MAX_PITUUS)
      Set(VALIDATION_OTSIKKO_LIIAN_PITKA)
    else
      Set.empty

  val kaannosAvainPattern: Regex = "[a-zA-Z0-9]+".r
  def validateLahettavaPalvelu(lahettavaPalvelu: Optional[String]): Set[String] =
    if (lahettavaPalvelu.isEmpty || lahettavaPalvelu.get.isEmpty)
      Set(VALIDATION_LAHETTAVA_PALVELU_TYHJA)
    else
      Some(Set.empty.asInstanceOf[Set[String]])
        .map(virheet =>
          if (lahettavaPalvelu.get.length > LAHETTAVAPALVELU_MAX_PITUUS)
            virheet.incl(VALIDATION_LAHETTAVA_PALVELU_LIIAN_PITKA) else virheet)
        .map(virheet =>
          if (!kaannosAvainPattern.matches(lahettavaPalvelu.get))
            virheet.incl(VALIDATION_LAHETTAVA_PALVELU_INVALID) else virheet).get

  val ophOidPattern: Regex = (VALIDATION_OPH_OID_PREFIX + "(\\.[0-9]+)+").r
  def validateLahettavanVirkailijanOID(oid: Optional[String]): Set[String] =
    if (oid.isEmpty)
      Set.empty
    else
      Some(Set.empty.asInstanceOf[Set[String]])
        .map(virheet =>
          if (!ophOidPattern.matches(oid.get()))
            virheet.incl(VALIDATION_LAHETTAJAN_OID_INVALID) else virheet)
        .map(virheet =>
          if (oid.get.length > VIRKAILIJAN_OID_MAX_PITUUS)
            virheet.incl(VALIDATION_LAHETTAJAN_OID_PITUUS) else virheet).get

  def validateLahettaja(lahettaja: Optional[Lahettaja]): Set[String] =
    if (lahettaja.isEmpty)
      Set(VALIDATION_LAHETTAJA_TYHJA)
    else
      Some(Set.empty.asInstanceOf[Set[String]])
        .map(virheet =>
          if (lahettaja.get.getNimi.isPresent && lahettaja.get.getNimi.get.length > LAHETTAJA_NIMI_MAX_PITUUS)
            virheet.incl(VALIDATION_LAHETTAJA_NIMI_LIIAN_PITKA) else virheet)
        .map(virheet =>
          if (lahettaja.get.getSahkopostiOsoite.isEmpty || lahettaja.get.getSahkopostiOsoite.get.length == 0)
            virheet.incl(VALIDATION_LAHETTAJAN_OSOITE_TYHJA)
          else if (!EmailValidator.getInstance(false).isValid(lahettaja.get.getSahkopostiOsoite.get))
            virheet.incl(VALIDATION_LAHETTAJAN_OSOITE_INVALID)
          else if (!lahettaja.get.getSahkopostiOsoite.get.endsWith(VALIDATION_OPH_DOMAIN))
            virheet.incl(VALIDATION_LAHETTAJAN_OSOITE_DOMAIN)
          else virheet).get

  def validateReplyTo(replyTo: Optional[String]): Set[String] =
    if (replyTo.isEmpty)
      Set.empty
    else if (!EmailValidator.getInstance(false).isValid(replyTo.get))
      Set(VALIDATION_REPLYTO_INVALID)
    else
      Set.empty

  def validatePrioriteetti(prioriteetti: Optional[String]): Set[String] =
    if (prioriteetti.isEmpty || (!prioriteetti.get.equals(LAHETYS_PRIORITEETTI_KORKEA) && !prioriteetti.get.equals(LAHETYS_PRIORITEETTI_NORMAALI)))
      Set(VALIDATION_PRIORITEETTI)
    else
      Set.empty

  def validateSailytysAika(sailytysAika: Optional[Integer]): Set[String] =
    if (sailytysAika.isEmpty)
      Set(VALIDATION_SAILYTYSAIKA_TYHJA)
    else if (sailytysAika.get < SAILYTYSAIKA_MIN_PITUUS || sailytysAika.get > SAILYTYSAIKA_MAX_PITUUS)
      Set(VALIDATION_SAILYTYSAIKA)
    else
      Set.empty

  def validateLahetys(lahetys: Lahetys): Set[String] =
    Set(
      validateOtsikko(lahetys.getOtsikko),
      validateLahettavaPalvelu(lahetys.getLahettavaPalvelu),
      validateLahettavanVirkailijanOID(lahetys.getLahettavanVirkailijanOid),
      validateLahettaja(lahetys.getLahettaja),
      validateReplyTo(lahetys.getReplyTo),
      validatePrioriteetti(lahetys.getPrioriteetti),
      validateSailytysAika(lahetys.getSailytysaika),
    ).flatten
  
end LahetysValidator

File: viestinvalitys-service/src/main/java/fi/vm/sade/viestinvalitys/validation/LahetysValidator.java
package fi.vm.sade.viestinvalitys.validation;

import fi.vm.sade.viestinvalitys.dto.Kontakti;
import fi.vm.sade.viestinvalitys.dto.LuoLahetysRequest;

import java.util.LinkedHashSet;
import java.util.Set;
import java.util.regex.Pattern;

import org.apache.commons.validator.routines.EmailValidator;

public final class LahetysValidator {

    private LahetysValidator() {}

    public static final int OTSIKKO_MAX_PITUUS = 255;
    public static final int LAHETTAVAPALVELU_MAX_PITUUS = 127;
    public static final int LAHETTAJA_NIMI_MAX_PITUUS = 64;
    public static final int VIRKAILIJAN_OID_MAX_PITUUS = 64;
    public static final int SAILYTYSAIKA_MIN_PITUUS = 1;
    public static final int SAILYTYSAIKA_MAX_PITUUS = 3650;
    public static final String LAHETYS_PRIORITEETTI_KORKEA = "korkea";
    public static final String LAHETYS_PRIORITEETTI_NORMAALI = "normaali";

    public static final String VALIDATION_OPH_OID_PREFIX = "1.2.246.562";
    public static final String VALIDATION_OPH_DOMAIN = "@opintopolku.fi";

    public static final String VALIDATION_OTSIKKO_TYHJA = "otsikko: Kenttä on pakollinen";
    public static final String VALIDATION_OTSIKKO_LIIAN_PITKA =
            "otsikko: Otsikko ei voi pidempi kuin " + OTSIKKO_MAX_PITUUS + " merkkiä";

    public static final String VALIDATION_LAHETTAVA_PALVELU_TYHJA = "lahettavaPalvelu: Kenttä on pakollinen";
    public static final String VALIDATION_LAHETTAVA_PALVELU_LIIAN_PITKA =
            "lahettavaPalvelu: Kentän pituus voi olla korkeintaan " + LAHETTAVAPALVELU_MAX_PITUUS + " merkkiä";

    public static final String VALIDATION_LAHETTAJAN_OID_INVALID =
            "lähettäjänOid: Oid ei ole validi (1.2.246.562-alkuinen) oph-oid";
    public static final String VALIDATION_LAHETTAJAN_OID_PITUUS =
            "lähettäjänOid-kentän suurin sallittu pituus on " + VIRKAILIJAN_OID_MAX_PITUUS + " merkkiä";

    public static final String VALIDATION_LAHETTAJA_TYHJA = "lähettäjä: Kenttä on pakollinen";
    public static final String VALIDATION_LAHETTAJA_NIMI_LIIAN_PITKA =
            "lähettäjä: nimi-kenttä voi maksimissaan olla " + LAHETTAJA_NIMI_MAX_PITUUS + " merkkiä pitkä";
    public static final String VALIDATION_LAHETTAJAN_OSOITE_TYHJA =
            "lähettäjä: Lähettäjän sähköpostiosoite -kenttä on pakollinen";
    public static final String VALIDATION_LAHETTAJAN_OSOITE_INVALID =
            "lähettäjä: Lähettäjän sähköpostiosoite ei ole validi sähköpostiosoite";
    public static final String VALIDATION_LAHETTAJAN_OSOITE_DOMAIN =
            "lähettäjä: Lähettäjän sähköpostiosoite ei ole opintopolku.fi -domainissa";

    public static final String VALIDATION_REPLYTO_INVALID = "replyTo: arvo ei ole validi sähköpostiosoite";

    public static final String VALIDATION_PRIORITEETTI =
            "prioriteetti: Prioriteetti täytyy olla joko \"" + LAHETYS_PRIORITEETTI_NORMAALI
                    + "\" tai \"" + LAHETYS_PRIORITEETTI_KORKEA + "\"";

    public static final String VALIDATION_SAILYTYSAIKA_TYHJA = "sailytysAika: Kenttä on pakollinen";
    public static final String VALIDATION_SAILYTYSAIKA =
            "sailytysAika: Säilytysajan tulee olla " + SAILYTYSAIKA_MIN_PITUUS + "-" + SAILYTYSAIKA_MAX_PITUUS + " päivää";

    private static final Pattern OPH_OID_PATTERN = Pattern.compile(VALIDATION_OPH_OID_PREFIX + "(\\.[0-9]+)+");

    public static Set<String> validateOtsikko(String otsikko) {
        Set<String> virheet = new LinkedHashSet<>();
        if (otsikko == null || otsikko.isEmpty()) {
            virheet.add(VALIDATION_OTSIKKO_TYHJA);
        } else if (otsikko.length() > OTSIKKO_MAX_PITUUS) {
            virheet.add(VALIDATION_OTSIKKO_LIIAN_PITKA);
        }
        return virheet;
    }

    public static Set<String> validateLahettavaPalvelu(String lahettavaPalvelu) {
        Set<String> virheet = new LinkedHashSet<>();
        if (lahettavaPalvelu == null || lahettavaPalvelu.isEmpty()) {
            virheet.add(VALIDATION_LAHETTAVA_PALVELU_TYHJA);
        } else if (lahettavaPalvelu.length() > LAHETTAVAPALVELU_MAX_PITUUS) {
            virheet.add(VALIDATION_LAHETTAVA_PALVELU_LIIAN_PITKA);
        }
        return virheet;
    }

    public static Set<String> validateLahettavanVirkailijanOID(String oid) {
        Set<String> virheet = new LinkedHashSet<>();
        if (oid == null) {
            return virheet;
        }
        if (!OPH_OID_PATTERN.matcher(oid).matches()) {
            virheet.add(VALIDATION_LAHETTAJAN_OID_INVALID);
        }
        if (oid.length() > VIRKAILIJAN_OID_MAX_PITUUS) {
            virheet.add(VALIDATION_LAHETTAJAN_OID_PITUUS);
        }
        return virheet;
    }

    public static Set<String> validateLahettaja(Kontakti lahettaja) {
        Set<String> virheet = new LinkedHashSet<>();
        if (lahettaja == null) {
            virheet.add(VALIDATION_LAHETTAJA_TYHJA);
            return virheet;
        }
        if (lahettaja.nimi() != null && lahettaja.nimi().length() > LAHETTAJA_NIMI_MAX_PITUUS) {
            virheet.add(VALIDATION_LAHETTAJA_NIMI_LIIAN_PITKA);
        }
        String osoite = lahettaja.sahkopostiOsoite();
        if (osoite == null || osoite.isEmpty()) {
            virheet.add(VALIDATION_LAHETTAJAN_OSOITE_TYHJA);
        } else if (!EmailValidator.getInstance(false).isValid(osoite)) {
            virheet.add(VALIDATION_LAHETTAJAN_OSOITE_INVALID);
        } else if (!osoite.endsWith(VALIDATION_OPH_DOMAIN)) {
            virheet.add(VALIDATION_LAHETTAJAN_OSOITE_DOMAIN);
        }
        return virheet;
    }

    public static Set<String> validateReplyTo(String replyTo) {
        Set<String> virheet = new LinkedHashSet<>();
        if (replyTo == null) {
            return virheet;
        }
        if (!EmailValidator.getInstance(false).isValid(replyTo)) {
            virheet.add(VALIDATION_REPLYTO_INVALID);
        }
        return virheet;
    }

    public static Set<String> validatePrioriteetti(String prioriteetti) {
        Set<String> virheet = new LinkedHashSet<>();
        if (prioriteetti == null
                || (!prioriteetti.equals(LAHETYS_PRIORITEETTI_KORKEA) && !prioriteetti.equals(LAHETYS_PRIORITEETTI_NORMAALI))) {
            virheet.add(VALIDATION_PRIORITEETTI);
        }
        return virheet;
    }

    public static Set<String> validateSailytysAika(Integer sailytysAika) {
        Set<String> virheet = new LinkedHashSet<>();
        if (sailytysAika == null) {
            virheet.add(VALIDATION_SAILYTYSAIKA_TYHJA);
        } else if (sailytysAika < SAILYTYSAIKA_MIN_PITUUS || sailytysAika > SAILYTYSAIKA_MAX_PITUUS) {
            virheet.add(VALIDATION_SAILYTYSAIKA);
        }
        return virheet;
    }

    public static Set<String> validateLahetys(LuoLahetysRequest lahetys) {
        Set<String> virheet = new LinkedHashSet<>();
        virheet.addAll(validateOtsikko(lahetys.otsikko()));
        virheet.addAll(validateLahettavaPalvelu(lahetys.lahettavaPalvelu()));
        virheet.addAll(validateLahettavanVirkailijanOID(lahetys.lahettavanVirkailijanOid()));
        virheet.addAll(validateLahettaja(lahetys.lahettaja()));
        virheet.addAll(validateReplyTo(lahetys.replyTo()));
        virheet.addAll(validatePrioriteetti(lahetys.prioriteetti()));
        virheet.addAll(validateSailytysAika(lahetys.sailytysaika()));
        return virheet;
    }
}
